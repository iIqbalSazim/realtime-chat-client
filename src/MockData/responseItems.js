export const mockLoginResponse = {
  data: {
    id: 123,
    username: "Test name",
    token: "mockToken",
  },
};

export const mockRegistrationResponse = {
  data: {
    id: 123,
    username: "Test name",
    token: "mockToken",
  },
};
