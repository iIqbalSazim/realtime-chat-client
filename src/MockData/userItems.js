export const mockUsers = [
  { id: 1, username: "Test User1" },
  { id: 2, username: "Test User2" },
];
