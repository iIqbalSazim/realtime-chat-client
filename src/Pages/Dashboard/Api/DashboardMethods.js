import { getRequest, postRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (receiverId) => {
  return await getRequest(`messages?chatting_with=${receiverId}`);
};

export const sendMessage = async (message) => {
  return await postRequest("messages", message);
};
