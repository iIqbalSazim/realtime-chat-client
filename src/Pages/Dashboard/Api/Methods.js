import { getRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (senderId, receiverId) => {
  return await getRequest(
    `messages?sender_id=${senderId}&receiver_id=${receiverId}`
  );
};
