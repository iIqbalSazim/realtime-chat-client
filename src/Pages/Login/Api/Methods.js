import { postRequest } from "../../../Config/Axios/AxiosConfig";

export const loginUser = async (user) => {
  return await postRequest("login", user);
};

export const revokeToken = async (token) => {
  return await postRequest("revoke_token", token);
};
