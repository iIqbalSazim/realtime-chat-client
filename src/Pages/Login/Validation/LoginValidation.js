export const LoginFormValidationSchema = {
  validateUsername: {
    required: { value: true, message: "Username is a required field." },
    minLength: {
      value: 3,
      message: "Username must be at least 3 characters",
    },
    maxLength: {
      value: 250,
      message: "Username must not exceed 250 characters",
    },
  },
  validatePassword: {
    required: { value: true, message: "Password is a required field." },
    minLength: {
      value: 8,
      message: "Password must be at least 8 characters.",
    },
    maxLength: {
      value: 250,
      message: "Password must not exceed 250 characters.",
    },
  },
};
