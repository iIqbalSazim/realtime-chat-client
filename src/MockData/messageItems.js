import { mockChattingWith, mockCurrentUser } from "./userItems";

export const mockMessages = [
  { id: 1, content: "Hello", sender_id: 1, receiver_id: 2 },
  { id: 2, content: "How are you?", sender_id: 1, receiver_id: 2 },
];

export const mockMessage = {
  id: 1,
  content: "Hello",
  sender_id: 1,
  receiver_id: 2,
};

export const mockIncomingMessage = {
  id: 1,
  content: "Hello",
  sender_id: 2,
  receiver_id: 1,
};

export const mockIncomingMessages = [
  { id: 1, content: "Hello", sender_id: 2, receiver_id: 1 },
  { id: 2, content: "How are you?", sender_id: 2, receiver_id: 1 },
];

export const mockNewMessage = {
  id: 1234,
  content: "This is a new message",
  sender_id: 1,
  receiver_id: 2,
};

export const mockChatMessages = [
  {
    direction: "outgoing",
    id: 1,
    message: "Hello",
    position: "single",
    sender_id: mockCurrentUser.id,
    receiver_id: mockChattingWith.id,
  },
  {
    direction: "incoming",
    id: 2,
    message: "Hey there",
    position: "single",
    sender_id: mockChattingWith.id,
    receiver_id: mockCurrentUser.id,
  },
  {
    direction: "outgoing",
    id: 3,
    message: "How are you?",
    position: "single",
    sender_id: mockCurrentUser.id,
    receiver_id: mockChattingWith.id,
  },
];
