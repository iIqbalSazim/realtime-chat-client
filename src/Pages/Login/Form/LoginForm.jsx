import { useState } from "react";
import { useForm } from "react-hook-form";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";

import { loginSuccess } from "../../../Stores/Actions/auth";

import { loginUser } from "../Api/Methods";
import { LoginFormValidationSchema } from "../Validation/LoginValidation";

const LoginForm = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const defaultValues = useState({
    username: "",
    password: "",
  });

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm({
    defaultValues,
  });

  const onSubmit = async (user) => {
    const response = await loginUser(user);

    if (response) {
      const newUser = {
        id: response.data.id,
        username: response.data.username,
      };

      const token = response.data.token;

      localStorage.setItem("token", token);

      dispatch(loginSuccess(newUser));

      alert("Logged in Successfully");

      navigate(`/${newUser.id}`);
    }

    reset({ ...defaultValues });
  };

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      className="flex flex-col gap-4 items-center justify-center w-[500px] px-10 py-12 border border-slate-200"
    >
      <h1 className="font-medium text-lg mb-5">LOGIN</h1>

      <div className="form-control w-full max-w-xs">
        <label className="label">
          <span className="label-text">Username</span>
        </label>

        <input
          type="text"
          placeholder="Your username"
          className="input input-bordered w-full max-w-xs"
          {...register("username", {
            ...LoginFormValidationSchema.validateUsername,
          })}
        />
        {errors.username && (
          <p className="text-red-500 text-xs my-1">{errors.username.message}</p>
        )}
      </div>

      <div className="form-control w-full max-w-xs">
        <label className="label">
          <span className="label-text">Password</span>
        </label>

        <input
          type="password"
          placeholder="Your password"
          className="input input-bordered w-full max-w-xs"
          {...register("password", {
            ...LoginFormValidationSchema.validatePassword,
          })}
        />
        {errors.password && (
          <p className="text-red-500 text-xs my-1">{errors.password.message}</p>
        )}
      </div>

      <button
        className="btn btn-md btn-secondary font-semibold mt-5"
        type="submit"
      >
        Submit
      </button>
    </form>
  );
};

export default LoginForm;
