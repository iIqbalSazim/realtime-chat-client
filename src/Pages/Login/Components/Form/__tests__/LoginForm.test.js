import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../../../../../../jest.setup";

import { mockLoginResponse } from "../../../../../MockData/responseItems";
import { LoginFormValidationSchema } from "../../../Validation/LoginValidation";
import * as apiMethods from "../../../Api/LoginMethods";
import LoginForm from "../LoginForm";

describe("LoginForm", () => {
  test("renders correctly", () => {
    const { container } = renderWithProviders(<LoginForm />);

    const headingElement = screen.getByRole("heading", {
      level: 1,
    });
    const usernameInput = screen.getByPlaceholderText(/Your username/i);
    const passwordInput = screen.getByPlaceholderText(/Your password/i);
    const submitButton = screen.getByRole("button", {
      name: "Submit",
    });

    expect(headingElement).toBeInTheDocument();
    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("call loginUser function with user as argument if form submitted with valid inputs", async () => {
    const loginUserSpy = jest.spyOn(apiMethods, "loginUser");

    loginUserSpy.mockResolvedValue(mockLoginResponse);
    renderWithProviders(<LoginForm />);

    const usernameInput = screen.getByPlaceholderText(/Your username/i);
    const passwordInput = screen.getByPlaceholderText(/Your password/i);
    const submitButton = screen.getByRole("button", {
      name: "Submit",
    });

    await userEvent.type(usernameInput, "Test name");
    await userEvent.type(passwordInput, "password");

    await userEvent.click(submitButton);

    expect(loginUserSpy).toHaveBeenCalledWith({
      username: "Test name",
      password: "password",
    });
  });

  test("renders required validation error if form submitted with empty inputs", async () => {
    const { container } = renderWithProviders(<LoginForm />);

    const submitButton = screen.getByRole("button", {
      name: "Submit",
    });

    await userEvent.click(submitButton);

    expect(
      screen.queryByText(
        LoginFormValidationSchema.validateUsername.required.message
      )
    ).toBeInTheDocument();
    expect(
      screen.queryByText(
        LoginFormValidationSchema.validatePassword.required.message
      )
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("renders minLength validation error if form submitted with short inputs", async () => {
    const { container } = renderWithProviders(<LoginForm />);

    const usernameInput = screen.getByPlaceholderText(/Your username/i);
    const passwordInput = screen.getByPlaceholderText(/Your password/i);
    const submitButton = screen.getByRole("button", {
      name: "Submit",
    });

    await userEvent.type(usernameInput, "12");
    await userEvent.type(passwordInput, "1234567");
    await userEvent.click(submitButton);

    expect(
      screen.queryByText(
        LoginFormValidationSchema.validateUsername.minLength.message
      )
    ).toBeInTheDocument();
    expect(
      screen.queryByText(
        LoginFormValidationSchema.validatePassword.minLength.message
      )
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("renders maxLength validation error if form submitted with long inputs", async () => {
    const { container } = renderWithProviders(<LoginForm />);

    const usernameInput = screen.getByPlaceholderText(/Your username/i);
    const passwordInput = screen.getByPlaceholderText(/Your password/i);
    const submitButton = screen.getByRole("button", {
      name: "Submit",
    });

    const longInput = "a".repeat(270);

    await userEvent.type(usernameInput, longInput);
    await userEvent.type(passwordInput, longInput);
    await userEvent.click(submitButton);

    expect(
      screen.queryByText(
        LoginFormValidationSchema.validateUsername.maxLength.message
      )
    ).toBeInTheDocument();
    expect(
      screen.queryByText(
        LoginFormValidationSchema.validatePassword.maxLength.message
      )
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
