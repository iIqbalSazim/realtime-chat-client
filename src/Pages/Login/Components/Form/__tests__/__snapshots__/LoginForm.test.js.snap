// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders correctly 1`] = `
<div>
  <form
    class="flex flex-col gap-4 items-center justify-center w-[500px] px-10 py-12 border border-slate-200"
  >
    <h1
      class="font-medium text-lg mb-5"
    >
      LOGIN
    </h1>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Username
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="username"
        placeholder="Your username"
        type="text"
      />
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Password
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="password"
        placeholder="Your password"
        type="password"
      />
    </div>
    <button
      class="btn btn-md btn-secondary font-semibold mt-5"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`LoginForm renders maxLength validation error if form submitted with long inputs 1`] = `
<div>
  <form
    class="flex flex-col gap-4 items-center justify-center w-[500px] px-10 py-12 border border-slate-200"
  >
    <h1
      class="font-medium text-lg mb-5"
    >
      LOGIN
    </h1>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Username
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="username"
        placeholder="Your username"
        type="text"
      />
      <p
        class="text-red-500 text-xs my-1"
      >
        Username must not exceed 250 characters
      </p>
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Password
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="password"
        placeholder="Your password"
        type="password"
      />
      <p
        class="text-red-500 text-xs my-1"
      >
        Password must not exceed 250 characters.
      </p>
    </div>
    <button
      class="btn btn-md btn-secondary font-semibold mt-5"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`LoginForm renders minLength validation error if form submitted with short inputs 1`] = `
<div>
  <form
    class="flex flex-col gap-4 items-center justify-center w-[500px] px-10 py-12 border border-slate-200"
  >
    <h1
      class="font-medium text-lg mb-5"
    >
      LOGIN
    </h1>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Username
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="username"
        placeholder="Your username"
        type="text"
      />
      <p
        class="text-red-500 text-xs my-1"
      >
        Username must be at least 3 characters
      </p>
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Password
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="password"
        placeholder="Your password"
        type="password"
      />
      <p
        class="text-red-500 text-xs my-1"
      >
        Password must be at least 8 characters.
      </p>
    </div>
    <button
      class="btn btn-md btn-secondary font-semibold mt-5"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`LoginForm renders required validation error if form submitted with empty inputs 1`] = `
<div>
  <form
    class="flex flex-col gap-4 items-center justify-center w-[500px] px-10 py-12 border border-slate-200"
  >
    <h1
      class="font-medium text-lg mb-5"
    >
      LOGIN
    </h1>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Username
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="username"
        placeholder="Your username"
        type="text"
      />
      <p
        class="text-red-500 text-xs my-1"
      >
        Username is a required field.
      </p>
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
      >
        <span
          class="label-text"
        >
          Password
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        name="password"
        placeholder="Your password"
        type="password"
      />
      <p
        class="text-red-500 text-xs my-1"
      >
        Password is a required field.
      </p>
    </div>
    <button
      class="btn btn-md btn-secondary font-semibold mt-5"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
